pairing/bn254/point.go

```diff
@@ -0,0 +1,786 @@
+package bn254
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"crypto/subtle"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/drand/kyber"
+	"github.com/drand/kyber/group/mod"
+	"golang.org/x/crypto/sha3"
+)
+
+var marshalPointID1 = [8]byte{'b', 'n', '2', '5', '4', '.', 'g', '1'}
+var marshalPointID2 = [8]byte{'b', 'n', '2', '5', '4', '.', 'g', '2'}
+var marshalPointIDT = [8]byte{'b', 'n', '2', '5', '4', '.', 'g', 't'}
+
+type pointG1 struct {
```

Comment:
Is there any reason why you needed to copy paste the whole gnark code instead of _using_ it ? 
It's gonna be difficult to get upstream changes (and vice versa) in this model.

Reply:
I spliced the existing [`bn256` package in this repo](https://github.com/drand/kyber/tree/master/pairing/bn256) with code from [geth](https://github.com/ethereum/go-ethereum/blob/master/crypto/bn256/cloudflare/). So it actually looks more like the local `bn256` code than geth as it has coupling between kyber interfaces and private/non-exported bn256 code. Both of these packages were originally vendored from https://github.com/cloudflare/bn256.

Some other differences:

- Groups have a domain separation tag now: https://github.com/drand/kyber/pull/55/files#diff-b94537aa3a0d5111b9d2e1833e5b031067c74e75d156dfad3570f9968237d008R23
- I changed the original [`Hash` function](https://github.com/drand/kyber/pull/55/files#diff-b94537aa3a0d5111b9d2e1833e5b031067c74e75d156dfad3570f9968237d008R212) from hash-and-pray to the constant-time Fouque-Tibouchi method. This function exists in this repo's `bn256` package but does not exist in the geth codebase.